src/app/modules/webhook/webhook.message.ts

```diff
@@ -0,0 +1,174 @@
+import { differenceInSeconds } from 'date-fns'
+import { Result } from 'neverthrow'
+
+import { createLoggerWithLabel } from '../../config/logger'
+
+import {
+  DUE_TIME_TOLERANCE_SECONDS,
+  QUEUE_MESSAGE_VERSION,
+} from './webhook.constants'
+import {
+  WebhookNoMoreRetriesError,
+  WebhookQueueMessageParsingError,
+} from './webhook.errors'
+import {
+  WebhookFailedQueueMessage,
+  webhookMessageSchema,
+  WebhookQueueMessageObject,
+  WebhookQueueMessagePrettified,
+} from './webhook.types'
+import { getNextAttempt, prettifyEpoch } from './webhook.utils'
+
+const logger = createLoggerWithLabel(module)
+
+/**
+ * Encapsulates a queue message for webhook retries.
+ */
+export class WebhookQueueMessage {
+  message: WebhookQueueMessageObject
+
+  constructor(message: WebhookQueueMessageObject) {
+    this.message = message
+  }
+
+  /**
+   * Converts a webhook queue message body into an encapsulated
+   * class instance.
+   * @param body Raw body of webhook queue message
+   * @returns ok(encapsulated message) if message can be parsed successfully
+   * @returns err if message fails to be parsed
+   */
+  static deserialise(
+    body: string,
+  ): Result<WebhookQueueMessage, WebhookQueueMessageParsingError> {
+    return Result.fromThrowable(
+      () => JSON.parse(body) as unknown,
+      (error) => {
+        logger.error({
+          message: 'Unable to parse webhook queue message body',
+          meta: {
+            action: 'deserialise',
+            body,
+          },
+          error,
+        })
+        return new WebhookQueueMessageParsingError(error)
+      },
+    )()
+      .andThen((parsed) =>
+        Result.fromThrowable(
+          () => webhookMessageSchema.parse(parsed),
+          (error) => {
+            logger.error({
+              message: 'Webhook queue message body has wrong shape',
+              meta: {
+                action: 'deserialise',
+                body,
+              },
+              error,
+            })
+            return new WebhookQueueMessageParsingError(error)
+          },
+        )(),
+      )
+      .map((validated) => new WebhookQueueMessage(validated))
+  }
+
+  /**
+   * Initialises a webhook queue message which has not been
+   * retried as yet. This function succeeds as long as
+   * the retry policy allows for at least one retry.
+   * @param submissionId
+   * @returns ok(encapsulated message) if retry policy exists
+   * @returns err if the retry policy does not allow any retries
+   */
+  static fromSubmissionId(
+    submissionId: string,
+  ): Result<WebhookQueueMessage, WebhookNoMoreRetriesError> {
+    return getNextAttempt([]).map(
```

Comment:
should we add the parameter name to make it clearer (i was confused at why an array was given):

```ts
return getNextAttempt(/** previousAttempts= */ [])
```

or should [] be set as default?
or should there be another helper method `getFirstNextAttempt` that just calls `getNextAttempt([])`?